File: day016_binary_search/primary_binary_search/solution.py
from typing import List

class Solution:
    # Optimized: O(log N) Time, O(1) Space - Iterative
    def search_iterative(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        
        while left <= right:
            # Use left + (right - left) // 2 to prevent overflow
            mid = left + (right - left) // 2
            
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                left = mid + 1  # Target is in the right half
            else:
                right = mid - 1 # Target is in the left half
                
        return -1
    
    # Brute Force: O(N) Time
    def search_brute(self, nums: List[int], target: int) -> int:
        for i, num in enumerate(nums):
            if num == target:
                return i
        return -1

File: day016_binary_search/primary_binary_search/solution.cpp
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    // Optimized: O(log N) Time, O(1) Space - Iterative
    int search_iterative(const vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2; // Prevents overflow
            
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return -1;
    }
    
    // Brute Force: O(N) Time
    int search_brute(const vector<int>& nums, int target) {
        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] == target) {
                return i;
            }
        }
        return -1;
    }
};